Policy support in API (#141)
hashicorp/terraform-docs-common: website/docs/cloud-docs/api-docs/policy-sets.mdx

# Policy Sets API

> -> **Note:** Sentinel and OPA policies are available in the [Terraform Cloud **Team & Governance** tier](https://www.hashicorp.com/products/terraform/pricing), and OPA policies are in beta.

Reviewer: ```suggestion
-> **Note:** Sentinel and OPA policies are available in the [Terraform Cloud **Team & Governance** tier](https://www.hashicorp.com/products/terraform/pricing), and OPA policies are in beta. OPA policies are also not available in Terraform Enterprise.
```

> -> **Note:** Sentinel and OPA policies are available in the [Terraform Cloud **Team & Governance** tier](https://www.hashicorp.com/products/terraform/pricing), and OPA policies are in beta. OPA policies are also not available in Terraform Enterprise.

Reviewer: ```suggestion
-> **Note:** Sentinel and OPA policies are available in the [Terraform Cloud Team & Governance tier](https://www.hashicorp.com/products/terraform/pricing), and OPA policies are in beta. OPA policies are not available in Terraform Enterprise.
```

[Policy Enforcement](/cloud-docs/policy-enforcement) lets you use the policy-as-code frameworks Sentinel and Open Policy Agent (OPA) to apply policy checks to Terraform Cloud workspaces.

Policy sets let you group related policies by attributes like environment or region. You can then apply each policy set to one or more workspaces. For these workspaces, Terraform Cloud checks every Terraform plan against each policy in the set.

> You can create new policies with the [Policies API](/cloud-docs/api-docs/policies) and then manually add them to a policy set. You can also create a versioned policy set that you add to Terraform Cloud with the [Policy Set Versions API](#create-a-policy-set-version) or that Terraform Cloud retrieves from a version control system.
> 
> This page documents the API endpoints to create, read, update, and delete policy sets in an organization. To view and manage policies, use the [Policies API](/cloud-docs/api-docs/policies).

Reviewer: ```suggestion
This API provides endpoints to create, read, update, and delete policy sets in a Terraform Cloud organization. To view and manage individual policies, use the [Policies API](/cloud-docs/api-docs/policies). 

Many of these endpoints let you create policy sets from a designated repository in a Version Control System (VCS). For more information about how to configure a policy set VCS repository, refer to [Sentinel Policy Set VCS Repositories](/cloud-docs/policy-enforcement/sentinel/vcs) and [OPA Policy Set VCS Repositories](/cloud-docs/policy-enforcement/opa/vcs). 

Instead of connecting Terraform Cloud to a VCS repository, you can use the the [Policy Set Versions endpoint](#create-a-policy-set-version) to create an entire policy set from a `tar.gz` file. 
```

-> **Note:** Sentinel policies are a paid feature, available as part of the **Team & Governance** upgrade package. [Learn more about Terraform Cloud pricing here](https://www.hashicorp.com/products/terraform/pricing).

> [Policy as Code](/cloud-docs/policy-enforcement) is an embedded policy as code framework integrated with Terraform Cloud.

Reviewer: ```suggestion
[Policy Enforcement](/cloud-docs/policy-enforcement) lets you use the policy-as-code frameworks Sentinel and Open Policy Agent (OPA) to apply policy checks to Terraform Cloud workspaces.
```

> Policy sets are groups of policies that are applied together to related workspaces. By using policy sets, you can group your policies by attributes such as environment or region. Individual policies within a policy set will only be checked for workspaces that the policy set is attached to. Policy sets can group individual policies created via the [policies API](/cloud-docs/api-docs/policies), or act as versioned sets which are either sourced from a version control system (such as GitHub) or uploaded as a whole via the [policy set versions API](#create-a-policy-set-version).

Reviewer: I wanted to make this more active voice so it's easier to read...thoughts?
```suggestion
Policy sets let you group related policies by attributes like environment or region. You can then apply each policy set to one or more workspaces. For these workspaces, Terraform Cloud checks every Terraform plan against each policy in the set. 

You can create new policies with the [Policies API](/cloud-docs/api-docs/policies) and then manually add them to a policy set. You can also create a versioned policy set that you add to Terraform Cloud with the [Policy Set Versions API](#create-a-policy-set-version) or that Terraform Cloud retrieves from a version control system.
```

| `data.attributes.name`                        | string         |           | The name of the policy set. Can include letters, numbers, `-`, and `_`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                            |

> | `data.attributes.description`                 | string         | `null`    | A description of the set's purpose. This field supports Markdown and will be rendered in the Terraform Cloud UI.                                                                                                                                                                                                                                                                                                                                                                                                                                   |

Reviewer: ```suggestion
| `data.attributes.description`                 | string         | `null`    | Text describing the policy set's purpose. This field supports Markdown and appears in the Terraform Cloud UI.                                                                                                                                                                                                                                                                                                                                                                                                                                   |
```

> | `data.attributes.global`                      | boolean        | `false`   | Whether or not this policies in this set should be checked on all of the organization's workspaces, or only on workspaces the policy set is attached to.                                                                                                                                                                                                                                                                                                                                                                                           |

Reviewer: ```suggestion
| `data.attributes.global`                      | boolean        | `false`   | Whether Terraform Cloud should automatically apply this policy set to all of an organization's workspaces.                                                                                                                                                                                                                                                                                                                                                                                              |
```

> | `data.attributes.kind`                        | string         | `sentinel`| The kind of policy that will be created. Valid values are `sentinel` and `opa`                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |

Reviewer: ```suggestion
| `data.attributes.kind`                        | string         | `sentinel`| The policy-as-code framework associated with the policy. Valid values are `sentinel` and `opa`.                                                                                                                                                                                                                                                                                                                                                                                                                                                                     |
```

> | `data.attributes.overridable`                 | boolean        | `false`   | Whether or not this policy is overridable in case of failures. Only valid for OPA policies                                                                                                                                                                                                                                                                                                                                                                                                                                                         |

Reviewer: ```suggestion
| `data.attributes.overridable`                 | boolean        | `false`   | Whether or not users can override this policy when it fails during a run. Only valid for OPA policies.                                                                                                                                                                                                                                                                                                                                                                                                                                                         |
```

> | `data.attributes.vcs-repo`                    | object         | `null`    | VCS repository information. When present, the policies and configuration will be sourced from the specified VCS repository instead of being defined within Terraform Cloud. Note that this option and `policies` relationships are mutually exclusive and may not be used simultaneously.                                                                                                                                                                                                                                                          |

Reviewer: attribute or property?

```suggestion
| `data.attributes.vcs-repo`                    | object         | `null`    | VCS repository information. When present, Terraform Cloud sources the policies and configuration from the specified VCS repository. This attribute and `policies` relationships are mutually exclusive, and you cannot use them simultaneously.                                                                                                                                                                                                                                                          |
```

> | `data.attributes.vcs-repo.branch`             | string         | `null`    | The branch of the VCS repo. If empty, the VCS provider's default branch value will be used.                                                                                                                                                                                                                                                                                                                                                                                                                                                        |

Reviewer: ```suggestion
| `data.attributes.vcs-repo.branch`             | string         | `null`    | The branch of the VCS repository where Terraform Cloud should retrieve the policy set. If empty, Terraform Cloud uses the default branch.                                                                                                                                                                                                                                                                                                                                                                                                                                                        |
```

> | `data.attributes.vcs-repo.identifier`         | string         |           | The identifier of the VCS repository in the format `<namespace>/<repo>`. For example, on GitHub, this would be something like `hashicorp/my-policy-set`. The format for Azure DevOps is `<org>/<project>/_git/<repo>`.                                                                                                                                                                                                                                                                                                                             |

Reviewer: ```suggestion
| `data.attributes.vcs-repo.identifier`         | string         |           | The VCS repository identifier in the format `<namespace>/<repo>`. For example, `hashicorp/my-policy-set`. The format for Azure DevOps is `<org>/<project>/_git/<repo>`.                                                                                                                                                                                                                                                                                                                             |
```

> | `data.attributes.vcs-repo.oauth-token-id`     | string         |           | The OAuth Token ID to use to connect to the VCS host.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |

Reviewer: ```suggestion
| `data.attributes.vcs-repo.oauth-token-id`     | string         |           | The OAuth Token ID Terraform Cloud should use to connect to the VCS host.                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              |
```

> | `data.attributes.vcs-repo.ingress-submodules` | boolean        | `false`   | Determines whether repository submodules will be instantiated during the clone operation.                                                                                                                                                                                                                                                                                                                                                                                                                                                          |

Reviewer: The clone operation felt....very technical. Let me know if I misinterpreted what that means!

```suggestion
| `data.attributes.vcs-repo.ingress-submodules` | boolean        | `false`   | Whether Terraform Cloud should instantiate repository submodules when retrieving the policy set.                                                                                                                                                                                                                                                                                                                                                                                                                                                          |
```

> | `data.attributes.policies-path`               | string         | `null`    | The subdirectory of the attached VCS repository that contains the policies for this policy set. Files and directories outside of this sub-path will be ignored, and changes to those unrelated files won't cause the policy set to be updated. This option may only be specified when a VCS repo is present.                                                                                                                                                                                                                                       |

Reviewer: ```suggestion
| `data.attributes.policies-path`               | string         | `null`    | The VCS repository subdirectory that contains the policies for this policy set. Terraform Cloud ignores files and directories outside of this sub-path and does not update the policy set when those files change. This attribute is only valid when you specify a VCS repository for the policy set.                                                                                                                                                                                                                                      |
```

> | `data.relationships.workspaces.data[]`        | array\[object] | `[]`      | A list of resource identifier objects that defines which workspaces the new set will be attached to. These objects must contain `id` and `type` properties, and the `type` property must be `workspaces` (e.g. `{ "id": "ws-2HRvNs49EWPjDqT1", "type": "workspaces" }`). Obtain workspace IDs from the [workspace settings](/cloud-docs/workspaces/settings) or the [Show Workspace](/cloud-docs/api-docs/workspaces#show-workspace) endpoint. Individual workspaces cannot be attached to the policy set when `data.attributes.global` is `true`. |

Reviewer: ```suggestion
| `data.relationships.workspaces.data[]`        | array\[object] | `[]`      | A list of resource identifier objects that defines which workspaces are associated with the policy set. These objects must contain `id` and `type` properties, and the `type` property must be `workspaces`. For example, `{ "id": "ws-2HRvNs49EWPjDqT1", "type": "workspaces" }`. Obtain workspace IDs from the [workspace's settings page](/cloud-docs/workspaces/settings) or the [Show Workspace endpoint](/cloud-docs/api-docs/workspaces#show-workspace). You can only specify this attribute when `data.attributes.global` is `false`. |
```

> | `data.relationships.policies.data[]`          | array\[object] | `[]`      | A list of resource identifier objects that defines which policies will be members of the new set. These objects must contain `id` and `type` properties, and the `type` property must be `policies` (e.g. `{ "id": "pol-u3S5p2Uwk21keu1s", "type": "policies" }`).                                                                                                                                                                |

Reviewer: ```suggestion
| `data.relationships.policies.data[]`          | array\[object] | `[]`      | A list of resource identifier objects that defines which policies are members of the policy set. These objects must contain `id` and `type` properties, and the `type` property must be `policies`. For example, `{ "id": "pol-u3S5p2Uwk21keu1s", "type": "policies" }`.                                                                                                                                                                |
```

| Parameter           | Description                                                                                                                                                                                                                                                                                                                                      |
| ------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `filter[versioned]` | **Optional.** Allows filtering policy sets based on whether they are versioned (VCS-managed or API-managed), or use individual policy relationships. Accepts a boolean true/false value. A `true` value returns versioned sets, and a `false` value returns sets with individual policy relationships. If omitted, all policy sets are returned. |

> | `filter[kind]`      | **Optional.** If specified, restricts results to those with the matching policy kind value. Valid values are `sentinel` and `opa`                                                                                                                                                                                                                |                                                             |

Reviewer: ```suggestion
| `filter[kind]`      | **Optional.** If specified, restricts results to those with the matching policy kind value. Valid values are `sentinel` and `opa`.                                                                                                                                                                                                                |                                                             |
```
